### PR TITLE
RawBearer API

### PR DESCRIPTION
**Pull request by tdammers**
# Description

This provides a socket-like API for Snockets (backed by sockets / named pipes, or simulated in IOSim), reading/writing directly to/from raw memory buffers.

This will be needed to securely send KES keys over a network connection or local pipe, because we need to ensure that key data is never stored on disk, nor on the GHC heap, from where it might be swapped to disk.

For context, see also:

- https://github.com/input-output-hk/cardano-base/pull/255: KES Secure Forgetting
- https://github.com/input-output-hk/cardano-base/issues/312: KES Agent
- https://github.com/input-output-hk/cardano-base/pull/317: KES Agent prerequisites

# Checklist

- Branch
    - [X] Commit sequence broadly makes sense
    - [X] Commits have useful messages
    - [X] The documentation has been properly updated
    - [X] New tests are added if needed and existing tests are updated
    - [X] Any changes affecting Consensus packages must have an entry in the appropriate `changelog.d` directory created using [`scriv`](https://github.com/input-output-hk/scriv). If in doubt, see the [Consensus release process](../ouroboros-consensus/docs/ReleaseProcess.md).
    - (n/a) If this branch changes Network and has any consequences for downstream repositories or end users, said changes must be documented in [`interface-CHANGELOG.md`](../docs/interface-CHANGELOG.md)
    - [X] If serialization changes, user-facing consequences (e.g. replay from genesis) are confirmed to be intentional.
- Pull Request
    - [X] Self-reviewed the diff
    - [X] Useful pull request description at least containing the following information:
      - What does this PR change?
      - Why these changes were needed?
      - How does this affect downstream repositories and/or end-users?
      - Which ticket does this PR close (if any)? If it does, is it [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue)?
    - [X] Reviewer requested

### CODE REVIEW

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/src/Simulation/Network/Snocket.hs:550]:
```suggestion
-- can bypas the requirement of moving raw bytes directly between file
```

Comment (coot) [inline ouroboros-network-framework/src/Simulation/Network/Snocket.hs:550]:
```suggestion
-- can bypas the requirement of moving raw bytes directly between file
```

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/src/Simulation/Network/Snocket.hs:569]: I would avoid that, because this can be very noisy.  It's better to provide a tracer which the user can fill with `nullTracer` or something better for debugging.

That's another argument to abandon the `ToRawBearer` type class (we probably don't want to have that tracer for real fds).

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (coot) [inline ouroboros-network-framework/src/Simulation/Network/Snocket.hs:597]: Just a different code layout:
```suggestion
                      (lhs, rhs) <- if not (LBS.null bytesFromBuffer)
                        then do
                          say $ "Reading up to " ++ show size ++ " bytes from buffer"
                          return (LBS.take size64 bytesFromBuffer, LBS.drop size64 bytesFromBuffer)
                        else do
                          bytesRead <- acRead (connChannelLocal conn)
                          say $ "Received " ++ show (LBS.length $ LBS.take size64 bytesRead) ++ " or more bytes"
                          return (LBS.take size64 bytesRead, LBS.drop size64 bytesRead)
```

File: ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs:84]: I think that's simpler, but I don't mind `catchJust`.
```suggestion
        catch (removeFile name)
              (\e -> if isDoesNotExistErrorType (ioeGetErrorType e)
                     then return ()
                     else throwIO e)

```

File: ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs:109]:
```suggestion
newtype TestError = TestError String
  deriving Show
```

File: ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs:167]: Will this be safer?
```suggestion
                        let go _ n | n <= 0 = do
```

File: ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs
Comment (coot) [inline ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs:141]: Is this safer?
```suggestion
                      let go _ n | n <= 0 = do
```

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (tdammers) [reply ouroboros-network-framework/src/Simulation/Network/Snocket.hs:550]: I don't think so. https://www.thefreedictionary.com/bypass

File: ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs
Comment (tdammers) [reply ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs:39]: Sure, we can express the typeclass instance here as a plain function, and then express the typeclass instance in terms of that, like we do for Win32.

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (tdammers) [reply ouroboros-network-framework/src/Simulation/Network/Snocket.hs:569]: Ah, but wait - this is the instance for the simulated snocket, which can only be used in `IOSim`. The "real" instances do not demand `MonadSay`, and they do not emit any trace logging at all. And `IOSim` itself already routes `say` through a `Tracer`, so this is actually working just fine - we get traces when running in `IOSim`, but no performance overhead when running in `IO`.

File: ouroboros-network-framework/changelog.d/20230224_094922_tdammers_raw_bearer.rst
Comment (tdammers) [reply ouroboros-network-framework/changelog.d/20230224_094922_tdammers_raw_bearer.rst]: Fixed.

File: ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs
Comment (coot) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs:189]: So what about:
```hs
let go _ n | n < 0 = error "ups, I was wrong ;)"
```
:wink:

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (tdammers) [reply ouroboros-network-framework/src/Simulation/Network/Snocket.hs:569]: Indeed, I've also thought of adding it to `Snocket`, but that would add a lot of noise to all snocket code, despite being a fairly niche feature - I expect that the only practical application of this will be for the KES Agent, where we need to send private keys over network sockets without ever leaking them to disk (including swap); anything else can just use the existing networking features and plain old serialization through intermediate values on the Haskell heap.

Your idea of injecting the tracer through `makeRawFDBearer` won't work, because the test code needs to run against both simulated and real snockets; since it's generalized over the `fd` type, it doesn't know to call `makeRawFDBearer`, so unless we give up on the generalization (which means we can no longer test the actual identical code against `IOSim` unchanged, which is the whole point of this), the only ways I can see for this to work would be:

1. Extend the `send` and `recv` methods to accept a `Tracer` as an extra argument. Downside is that we'll be passing tracers to production instances, which will then silently ignore it - a lot of noise, potentially misleading, and there might even be performance overhead to it as well.
2. Pass a function `fd -> m (RawBearer m)` around as an argument; we can now inject the tracer by closing that function over it. This is essentially the same as the typeclass approach, except that we're passing the typeclass dictionary (or the single method, in this case) around explicitly.
3. Same as 2., but make that function part of the `Snocket` data structure.

I'm not a huge fan of either, but if we really feel the need to change the tracer, then option 2 would be my preference.

File: ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs
Comment (tdammers) [reply ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs:46]: I've actually gotten rid of the typeclass after all.

File: ouroboros-network-framework/src/Simulation/Network/Snocket.hs
Comment (tdammers) [reply ouroboros-network-framework/src/Simulation/Network/Snocket.hs]: OK, ran with it and implemented a better tracing solution:
- got rid of the typeclass after all
- introduced `MakeRawBearer` type (analogous to `MakeBearer`)
- stringly-typed debug logging calls inside the simulated socket's RawBearer now use a structured trace message type, `FDRawBearerTrace`
- when making a `MakeRawBearer` for simulated sockets, caller must pass a `Tracer m FDRawBearerTrace`

File: ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs
Comment (tdammers) [reply ouroboros-network-framework/test/Test/Ouroboros/Network/RawBearer.hs:166]: As above.

**Review by coot — APPROVED** [review]
LGTM.

**Comment by coot** [comment]
It can be merged once CI passes, we should merge it after #4571.

File: ouroboros-network-framework/CHANGELOG.md
Comment (coot) [inline ouroboros-network-framework/CHANGELOG.md:12]: You should add this to a new section which starts with:
```
## next version

### Non-breaking changes
...
```

**Comment by coot** [comment]
There's a conflict in `CHANGELOG.md`; Could you squash / clean the branch a bit before merging?

**Review by coot — CHANGES_REQUESTED** [review]

**Comment by coot** [comment]
@tdammers do we still need this PR?

File: ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs:46]: The haddock needs to be update, since there's not `ToRawBearer` type class.

File: ouroboros-network-framework/test/Main.hs
Comment (coot) [inline ouroboros-network-framework/test/Main.hs:28]: I think you want to add the tests to `ouroboros-network-framework:io-tests`; the `Main.hs` might be an artefact of a rebase error.

**Review by coot — CHANGES_REQUESTED** [review]

File: ouroboros-network-framework/test/Main.hs
Comment (tdammers) [reply ouroboros-network-framework/test/Main.hs:28]: Refactored things into the two test suites.

**Comment by coot** [comment]
@tdammers have you pushed your changes?

**Comment by tdammers** [comment]
Everything pushed now.

File: ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs
Comment (coot) [reply ouroboros-network-framework/src/Ouroboros/Network/RawBearer.hs:46]: ^ ping

Also the CHANGELOG mentions the type class.

**Review by coot — APPROVED** [review]
LGTM

**Pull request merged**
